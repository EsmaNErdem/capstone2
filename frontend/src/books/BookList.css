/* BookList.css */

.BookList-list {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
    margin-top: 20px; 
  }
  
  @media (min-width: 768px) {
    .BookList-list {
      justify-content: space-between;
    }
  }
  
  @media (min-width: 992px) {
    .BookList-list {
      justify-content: space-between;
    }
  }
  
  @media (min-width: 1200px) {
    .BookList-list {
      justify-content: space-between;
    }
  }
  