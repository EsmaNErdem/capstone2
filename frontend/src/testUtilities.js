import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
