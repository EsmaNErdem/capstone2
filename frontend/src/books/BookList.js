import React, { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import BookClubApi from "../api";
import Loading from "../utilities/Loading";
import BookSearchForm from "./BookSearchForm";
import BookCard from "./BookCard";
import Alert from "../utilities/Alert"
import InfiniteScroll from "react-infinite-scroll-component";
import "./BookList.css"

/**
 * Displays a list of books and search box.
 * The `BookList` component displays a list of books fetched from the Google Books API,
 * along with database reviews and likes. It supports both infinite scrolling for loading
 * more books and a search feature to filter books based on user input.
 * 
 * - API call loads books data when the component mounts with getBookList and when the search box is submitted redirects to /books/search.
 * - This component is designed for the "/books" route.
 * - Utilizes BookCard and BookSearchForm components and called by Routes
 * 
 * - Routes ==> BookCard, BookSearchForm
 */
const BookList = () => {
  console.debug("BookList");

  const navigate = useNavigate();
  const [loading, setLoading] = useState(true)
  const [books, setBooks] = useState([]);
  const [indexList, setIndexList] = useState(20);
  const [hasMore, setHasMore] = useState(true);
  const [error, setError] = useState(null);

  /**
   * Fetches the initial batch of books from the Google Books API and sets the state.
   */  
  useEffect(function getBooksOnMount() {
    console.debug("BookList useEffect getBooksOnMount");

    // Make an API call to get first batch of books data from backend when there is no searchDatan submitted
    const getFirstList =  async () => {
      setError(null);
      try {
        const booksApi = await BookClubApi.getBookList(0);
        setBooks(booksApi);
      } catch (e) {
        console.error("BookList useEffect API call data loading error:", e);
        setError("An error occurred while fetching books.");
      }
      
      setLoading(false)
    }

    // set loading to true while async getCurrentUser runs; once the
    // data is fetched (or even if an error happens!), this will be set back
    // to true to control the spinner.
    setLoading(true)
    getFirstList();
  }, []);


  // Gets more books list data as user infinite scrolls
  const getBookList = async () => {
    setError(null);
    try {
        const booksApi = await BookClubApi.getBookList(indexList);
        booksApi.length > 0 ? setHasMore(true) : setHasMore(false);
        setBooks(b => [...b, ...booksApi]);
    } catch (e) {
        console.error("BookList API call data loading error:", e);
        setError("An error occurred while fetching books.");
    }

    setLoading(false)
    setIndexList(index => index + 20)
  }

  const searchBookData = (searchData) => {
    let queryTerms = ""
    queryTerms = searchData.terms.title ? `&title=${searchData.terms.title}` : ""
    queryTerms += searchData.terms.author ? `&author=${searchData.terms.author}` : ""
    queryTerms += searchData.terms.publisher ? `&publisher=${searchData.terms.publisher}` : ""
    queryTerms += searchData.terms.subject ? `&title=${searchData.terms.subject}` : ""

    navigate(`/books/search?search=${searchData.search}${queryTerms}`);
  }

  if (loading) return <Loading />;

  return (
  <InfiniteScroll
    dataLength={books.length}
    next={getBookList}
    hasMore={hasMore}
    loader={<Loading />}
  >
      <div className="BookList col-md-8 offset-md-2">
        <BookSearchForm searchFor={searchBookData} />
        {books.length
            ? (
                <div className="BookList-list">
                  {books.map(({ id, title, author, description, publisher, category, cover, bookLikeCount, reviews }) => (
                      <BookCard 
                          key={id}
                          id={id}
                          title={title}
                          author={author}
                          description={description}
                          publisher={publisher}
                          category={category}
                          cover={cover}
                          bookLikeCount={+bookLikeCount}
                          reviews={reviews}
                      />
                  ))}
                </div>
            ) : (
                <>
                  <p className="lead">Sorry, no results were found!</p>
                  {error ? <Alert type="danger" messages={[error]} />: null}
                </>
            )}
      </div>
    </InfiniteScroll>
  );
}

export default BookList;