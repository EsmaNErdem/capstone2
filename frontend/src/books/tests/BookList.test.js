import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import BookList from '../BookList';
import BookClubApi from '../../api';
import * as router from 'react-router'

const navigate = jest.fn()

const mockBooks = [
  {
    id: 1,
    title: 'Book 1',
    author: 'Author 1',
    description: 'Description 1',
    publisher: 'Publisher 1',
    category: 'Category 1',
    cover: 'Cover 1',
    bookLikeCount: 5,
    reviews: ["Review1"],
  },
  {
    id: 2,
    title: 'Book 2',
    author: 'Author 2',
    description: 'Description 2',
    publisher: 'Publisher 2',
    category: 'Category 2',
    cover: 'Cover 2',
    bookLikeCount: 10,
    reviews: [],
  },
];

beforeEach(() => {
  jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate)
})

jest.mock("../../api", () => ({
  __esModule: true,
  default: {
    getBookList: jest.fn(),
  },
}));

afterEach(() => {
  jest.clearAllMocks();
});

test('renders without crashing', async () => {
  BookClubApi.getBookList.mockImplementationOnce(() => (mockBooks));

  await act(async () => {
    render(
      <MemoryRouter>
        <UserProvider>
          <BookList />
        </UserProvider>
      </MemoryRouter>
    );
  });

});

test('it renders and matches with snapshot', async () => {
  BookClubApi.getBookList.mockImplementationOnce(() => (mockBooks));

  await act(async () => {
    ({ asFragment } = render(
      <MemoryRouter>
        <UserProvider>
          <BookList />
        </UserProvider>
      </MemoryRouter>
    ));
  });

  expect(asFragment()).toMatchSnapshot();
});

test('renders mock data', async () => {
  BookClubApi.getBookList.mockImplementationOnce(() => (mockBooks));

  await act(async () => {
      ({ getByText } = render(
        <MemoryRouter>
          <UserProvider>
            <BookList />
          </UserProvider>
      </MemoryRouter>
    ));
  });
  
  expect(getByText("Description 1")).toBeInTheDocument()
});

test('search redirects to the correct route', async () => {
  BookClubApi.getBookList.mockImplementationOnce(() => (mockBooks));

  await act(async () => {
    ({ getByPlaceholderText } = render(
      <MemoryRouter>
        <UserProvider>
          <BookList />
        </UserProvider>
    </MemoryRouter>
  ));
  });

  const searchInput = getByPlaceholderText('Book search');
  fireEvent.change(searchInput, { target: { value: 'Harry' } });
  fireEvent.submit(searchInput);
  expect(navigate).toHaveBeenCalledWith('/books/search?search=Harry');
});

test('when Api call throws error, handles error', async () => {
  BookClubApi.getBookList.mockImplementationOnce(() => {
    throw new Error("API error");
  })
  jest.spyOn(console, 'error').mockImplementation(() => {}); // Mocking console.error to avoid error message in test output

  await act(async () => {
    ({ getByText } = render(
      <MemoryRouter>
        <UserProvider>
          <BookList />
        </UserProvider>
      </MemoryRouter>
    ));
  });

  await waitFor(() => {});
  expect(getByText("An error occurred while fetching books.")).toBeInTheDocument()
  console.error.mockRestore(); 
});

