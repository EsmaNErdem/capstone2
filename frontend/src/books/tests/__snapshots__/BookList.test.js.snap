// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: auto;"
    >
      <div
        class="BookList col-md-8 offset-md-2"
      >
        <div
          class="SearchBox "
        >
          <form
            class="SearchForm"
          >
            <div
              class="SearchInputContainer"
            >
              <input
                placeholder="Book search"
                value=""
              />
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-magnifying-glass SubmitButton-inside"
                data-icon="magnifying-glass"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <button
              class="ToggleButton-large"
              data-testid="filter-toggle"
              title="Toggle Advanced Search"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-filter "
                data-icon="filter"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.9 54.9C10.5 40.9 24.5 32 40 32H472c15.5 0 29.5 8.9 36.1 22.9s4.6 30.5-5.2 42.5L320 320.9V448c0 12.1-6.8 23.2-17.7 28.6s-23.8 4.3-33.5-3l-64-48c-8.1-6-12.8-15.5-12.8-25.6V320.9L9 97.3C-.7 85.4-2.8 68.8 3.9 54.9z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </form>
        </div>
        <div
          class="BookList-list"
        >
          <div
            class="BookCard"
          >
            <a
              class="Title"
              href="/books/1"
            >
              <img
                alt="Book 1"
                class="CardImage"
                src="Cover 1"
              />
            </a>
            <div
              class="CardContent"
            >
              <a
                class="Title"
                data-testid="book-title-link"
                href="/books/1"
              >
                <h2
                  data-testid="book-title"
                >
                  Book 1
                </h2>
              </a>
              <h3
                data-testid="book-author"
              >
                by Author 1
              </h3>
              <p>
                Description 1
              </p>
              <div
                class="CardFooter"
              >
                <div
                  style="display: flex; align-items: center;"
                >
                  <span
                    data-testid="review-count"
                  >
                    1
                  </span>
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                    data-testid="review-button"
                    style="color: yellowgreen;"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="CommentIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM18 14H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <div
                  style="display: flex; align-items: center;"
                >
                  <span
                    data-testid="book-card-like-count"
                  >
                    5
                  </span>
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                    data-testid="book-card-like-button"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="FavoriteIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
            <div
              class="PrivateSwipeArea-root css-x15wq9"
              style="width: 20px;"
            />
          </div>
          <div
            class="BookCard"
          >
            <a
              class="Title"
              href="/books/2"
            >
              <img
                alt="Book 2"
                class="CardImage"
                src="Cover 2"
              />
            </a>
            <div
              class="CardContent"
            >
              <a
                class="Title"
                data-testid="book-title-link"
                href="/books/2"
              >
                <h2
                  data-testid="book-title"
                >
                  Book 2
                </h2>
              </a>
              <h3
                data-testid="book-author"
              >
                by Author 2
              </h3>
              <p>
                Description 2
              </p>
              <div
                class="CardFooter"
              >
                <div
                  style="display: flex; align-items: center;"
                >
                  <span
                    data-testid="review-count"
                  >
                    0
                  </span>
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                    data-testid="review-button"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="CommentIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM18 14H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <div
                  style="display: flex; align-items: center;"
                >
                  <span
                    data-testid="book-card-like-count"
                  >
                    10
                  </span>
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                    data-testid="book-card-like-button"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="FavoriteIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
            <div
              class="PrivateSwipeArea-root css-x15wq9"
              style="width: 20px;"
            />
          </div>
        </div>
      </div>
      <div
        class="Loading"
      >
        Loading ...
      </div>
    </div>
  </div>
</DocumentFragment>
`;
